consensus-types/blocks/proto.go

```diff
@@ -0,0 +1,464 @@
+		}
+		return &eth.BeaconBlockBellatrix{
+			Slot:          b.slot,
+			ProposerIndex: b.proposerIndex,
+			ParentRoot:    b.parentRoot,
+			StateRoot:     b.stateRoot,
+			Body:          body,
+		}, nil
+	case version.BellatrixBlind:
+		body, ok := bodyMessage.(*eth.BlindedBeaconBlockBodyBellatrix)
+		if !ok {
+			return nil, errors.Wrap(err, incorrectBodyVersion)
+		}
+		return &eth.BlindedBeaconBlockBellatrix{
+			Slot:          b.slot,
+			ProposerIndex: b.proposerIndex,
+			ParentRoot:    b.parentRoot,
+			StateRoot:     b.stateRoot,
+			Body:          body,
+		}, nil
+	default:
+		return nil, errors.New("unsupported beacon block version")
+	}
+}
+
+// Proto returns the underlying protobuf beacon block body.
+func (b *BeaconBlockBody) Proto() (proto.Message, error) {
+	if b == nil {
+		return nil, errNilBody
+	}
+
+	switch b.version {
+	case version.Phase0:
+		return &eth.BeaconBlockBody{
+			RandaoReveal:      b.randaoReveal,
+			Eth1Data:          b.eth1Data,
+			Graffiti:          b.graffiti,
+			ProposerSlashings: b.proposerSlashings,
+			AttesterSlashings: b.attesterSlashings,
+			Attestations:      b.attestations,
+			Deposits:          b.deposits,
+			VoluntaryExits:    b.voluntaryExits,
+		}, nil
+	case version.Altair:
+		return &eth.BeaconBlockBodyAltair{
+			RandaoReveal:      b.randaoReveal,
+			Eth1Data:          b.eth1Data,
+			Graffiti:          b.graffiti,
+			ProposerSlashings: b.proposerSlashings,
+			AttesterSlashings: b.attesterSlashings,
+			Attestations:      b.attestations,
+			Deposits:          b.deposits,
+			VoluntaryExits:    b.voluntaryExits,
+			SyncAggregate:     b.syncAggregate,
+		}, nil
+	case version.Bellatrix:
+		return &eth.BeaconBlockBodyBellatrix{
+			RandaoReveal:      b.randaoReveal,
+			Eth1Data:          b.eth1Data,
+			Graffiti:          b.graffiti,
+			ProposerSlashings: b.proposerSlashings,
+			AttesterSlashings: b.attesterSlashings,
+			Attestations:      b.attestations,
+			Deposits:          b.deposits,
+			VoluntaryExits:    b.voluntaryExits,
+			SyncAggregate:     b.syncAggregate,
+			ExecutionPayload:  b.executionPayload,
+		}, nil
+	case version.BellatrixBlind:
+		return &eth.BlindedBeaconBlockBodyBellatrix{
+			RandaoReveal:           b.randaoReveal,
+			Eth1Data:               b.eth1Data,
+			Graffiti:               b.graffiti,
+			ProposerSlashings:      b.proposerSlashings,
+			AttesterSlashings:      b.attesterSlashings,
+			Attestations:           b.attestations,
+			Deposits:               b.deposits,
+			VoluntaryExits:         b.voluntaryExits,
+			SyncAggregate:          b.syncAggregate,
+			ExecutionPayloadHeader: b.executionPayloadHeader,
+		}, nil
+	default:
+		return nil, errors.New("unsupported beacon block body version")
+	}
+}
+
+func initSignedBlockFromProtoPhase0(pb *eth.SignedBeaconBlock) (*SignedBeaconBlock, error) {
+	if pb == nil {
+		return nil, errNilBlock
+	}
+
+	pb, ok := proto.Clone(pb).(*eth.SignedBeaconBlock)
```

Comment:
what is the reason for doing a clone here, this is going to make unmarshalling blocks expensive on memory and cpu